File: src/graph_config.py
from __future__ import print_function, division
from math import sqrt
from itertools import chain, product
from functools import reduce
import operator

class GraphConfig(object):
    def __init__(self, dim, r=-1):
        assert(type(dim) in (int, tuple, list))
        if type(dim) is int: dim = [dim]
        self._dim = dim
        self._const = list(map(
            lambda i: reduce(operator.mul, dim[i+1:], 1),
            range(len(dim))
        ))
        self._n = self._const[0] * dim[0]
        self._r = r

    def get_r(self):
        return self._r

    def set_r(self, r):
        self._r = r

    def get_dim(self):
        return self._dim

    def get_n(self):
        return self._n

    def vid(self, dim):
        assert type(dim) in (int, tuple, list)
        if type(dim) is int: dim = (dim,)
        if len(dim) == len(self._dim) and\
           all(map(lambda x, y: x >= 0 and x < y, dim, self._dim)):
            return sum(map(
                lambda x, y: x * y, dim, self._const
            ))
        else:
            return -1

    def loc(self, vid):
        return tuple(map(
            lambda c, d: vid // c % d,
            self._const, self._dim
        ))

    def dist(self, u, v):
        loc_u, loc_v = self.loc(u), self.loc(v)
        return sum(map(lambda a, b: abs(a-b), loc_u, loc_v))

    def are_neighbors(self, u, v):
        return u != v and self.dist(u, v) <= self._r if self._r >= 0 else\
                    True

    def neighbors(self, u):
        if(self._r < 0):
            return chain(range(u), range(u+1, self._n))
        loc = self.loc(u)
        return filter(
            lambda v: v >= 0,
            map(
                lambda d: self.vid(tuple(map(sum, zip(loc, d)))),
                filter(
                    lambda d: sum(map(abs, d)) <= self._r and\
                    any(map(lambda e: e != 0, d)),
                    product(*map(
                        lambda d: range(-self._r, self._r+1),
                        self._dim)
                    )
                )
            )
        )

def main():
    """
      0  1  2  3  4  5  6  7
      8  9 10 11 12 13 14 15
     16 17 18 19 20 21 22 23
     24 25 26 27 28 29 30 31
    """
    conf1 = GraphConfig((4, 8), 2)
    print(conf1.vid((1, 2)))
    print(conf1.loc(19))
    print(conf1.dist(10, 20))
    print(conf1.are_neighbors(11, 17))
    print(tuple(conf1.neighbors(11)))

if __name__ == '__main__':
    main()

File: src/graph_config.cpp
#include "graph_config.hpp"

#include <algorithm>
#include <cassert>
#include <cmath>
#include <numeric>
#include <utility>

#include <iostream>

namespace gg {

	GraphConfig::GraphConfig(const std::vector<int> &dim, const int &r) :
		_dim(dim), _r(r) {
		_const = std::vector<int>(dim.size());
		_const[dim.size()-1] = 1;
		for(int i = dim.size()-2; i >= 0; i--)
			_const[i] = _const[i+1] * dim[i+1];
		_n = _const[0] * dim[0];
		_r = r;
	}

	Vertex GraphConfig::vid(const std::vector<int> &loc) const {
		assert(_dim.size() == loc.size());
		for(std::size_t i = 0; i < loc.size(); i++)
			if(loc[i] < 0 || loc[i] >= _dim[i])
				return -1;
		Vertex _vid = 0;
		for(std::size_t i = 0; i < loc.size(); i++)
			_vid += loc[i] * _const[i];
		return _vid;
	}

	std::vector<int> GraphConfig::loc(const Vertex &vid) const {
		std::vector<int> _loc(_dim.size());
		for(std::size_t i = 0; i < _dim.size(); i++)
			_loc[i] = vid / _const[i] % _dim[i];
		return _loc;
	}

	int GraphConfig::dist(const Vertex &u, const Vertex &v) const {
		std::vector<int> loc_u = loc(u);
		std::vector<int> loc_v = loc(v);
		int _dist = 0;
		for(std::size_t i = 0; i < loc_u.size(); i++)
			_dist += abs(loc_u[i] - loc_v[i]);
		return _dist;
	}

	bool GraphConfig::are_neighbors(const Vertex &u, const Vertex &v) const {
		if(_r < 0) return true;
		else return u != v && dist(u, v) <= _r;
	}

	std::vector<Vertex> GraphConfig::neighbors(const Vertex &v) const {
		std::vector<Vertex> nei;
    if(_dim.size() == 1 || _r < 0) {
      nei.resize(_n - 1);
      std::generate(nei.begin(), nei.begin()+(v-1),
                    [n = 0] () mutable { return n++; });
      std::generate(nei.begin()+(v+1), nei.end(),
                    [n = v+1] () mutable { return n++; });
    } else {
      std::vector<int> minidim(_dim.size());
      std::fill(minidim.begin(), minidim.end(), 2*_r+1);
      GraphConfig miniconf(minidim);
      int maxidx = std::accumulate(minidim.begin(), minidim.end(), 1,
                                   std::multiplies<int>());
      std::vector<int> v_loc = this->loc(v);
      for(int di = 1; di <= maxidx; di++) {
        std::vector<int> loc(_dim.size());
        bool valid_loc = true;
        std::vector<int> diff = miniconf.loc(di);
        std::transform(diff.begin(), diff.end(), diff.begin(),
                       [r = _r](int d) -> int { return d - r; });
        int sumdiff = 0;
        for(size_t i = 0; i < loc.size(); i++) {
          loc[i] = v_loc[i] + diff[i];
          sumdiff += abs(diff[i]);
          std::cout << loc[i] << " ";
          if(loc[i] < 0 || loc[i] >= _dim[i] || sumdiff > _r) {
            valid_loc = false;
            break;
          }
        }
        std::cout << std::endl;
        if(!valid_loc || sumdiff == 0) continue;
        Vertex w = vid(loc);
        if(w >= 0) nei.push_back(w);
      }
      std::cout << std::endl;
    }
		return nei;
	}

}
